File: algorithms/1_easy/P/01175_prime_arrangements/solution.c
#define MOD 1000000007

static inline bool is_prime(const int n) {
    if (n <= 1) {
        return false;
    }
    const int sq = (int)sqrt(n);
    for (int i = 2; i <= sq; i += 1) {
        if ((n % i) == 0) {
            return false;
        }
    }
    return true;
}

static inline int64_t fact_with_mod(const int64_t n) {
    int64_t res = 1;
    for (int64_t i = 2; i <= n; i += 1) {
        res = (res * i) % MOD;
    }
    return res;
}

int numPrimeArrangements(int n) {
    int64_t p = 0;
    int64_t o = 0;
    for (int i = 1; i <= n; i += 1) {
        if (is_prime(i)) {
            p += 1;
        } else {
            o += 1;
        }
    }

    const int64_t res = fact_with_mod(p) * fact_with_mod(o);

    return (int)(res % MOD);
}

File: algorithms/1_easy/P/01175_prime_arrangements/solution.rs
impl Solution {
    const MOD: i64 = 1_000_000_007;

    #[inline(always)]
    fn is_prime(n: i32) -> bool {
        if n <= 1 {
            return false;
        }
        let sq = n.isqrt();
        for i in 2..=sq {
            if n % i == 0 {
                return false;
            }
        }
        true
    }

    #[inline(always)]
    fn fact_with_mod(n: i64) -> i64 {
        let mut res = 1;
        for i in 2..=n {
            res = (res * i) % Self::MOD;
        }
        res
    }

    pub fn num_prime_arrangements(n: i32) -> i32 {
        let (mut p, mut o) = (0_i64, 0_i64);
        for i in 1..= n {
            if Self::is_prime(i) {
                p += 1;
            } else {
                o += 1;
            }
        }

        let res = Self::fact_with_mod(p) * Self::fact_with_mod(o);

        (res % Self::MOD) as i32
    }
}
